infra/base-images/base-builder/jcc/jcc2.go

```diff
@@ -0,0 +1,462 @@
+}
+
+func Compile(bin string, args []string) (int, string, string) {
+	// Generate ASTs f we define this ENV var.
+	if astDir := os.Getenv("JCC_GENERATE_AST_DIR"); astDir != "" {
+		GenerateASTs(bin, args, astDir)
+	}
+	// Run the actual command.
+	return ExecBuildCommand(bin, args)
+}
+
+func TryCompileAndFixHeadersOnce(bin string, cmd []string, filename string) (fixed, hasBrokenHeaders bool) {
+	retcode, _, err := Compile(bin, cmd)
+	if retcode == 0 {
+		fixed = true
+		hasBrokenHeaders = false
+		return
+	}
+	missingHeader, isMissing := ExtractMissingHeader(err)
+	if !isMissing {
+		fixed = false
+		hasBrokenHeaders = false
+		return
+	}
+
+	newHeaderPath, found := FindMissingHeader(missingHeader)
+	if !found {
+		fixed = false
+		hasBrokenHeaders = true
+		return false, true
+	}
+	ReplaceMissingHeaderInFile(filename, missingHeader, newHeaderPath)
+	return false, true
+}
+
+func FindMissingHeader(missingHeader string) (string, bool) {
+	envVar := "JCC_MISSING_HEADER_SEARCH_PATH"
+	var searchPath string
+	searchPath, exists := os.LookupEnv(envVar)
+	if !exists {
+		searchPath = "/src"
+	}
+	searchPath, _ = filepath.Abs(searchPath)
+	var headerLocation string
+	missingHeader = "/" + missingHeader
+	find := func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		if strings.HasSuffix(path, missingHeader) {
+			headerLocation = path
+			return nil
+		}
+		return nil
+	}
+	filepath.WalkDir(searchPath, find)
+	if headerLocation == "" {
+		return "", false
+	}
+	return headerLocation, true
+}
+
+func CppifyHeaderIncludesFromFile(srcFile string) error {
+	contentsBytes, err := ioutil.ReadFile(srcFile)
+	if err != nil {
+		return err
+	}
+	contents := string(contentsBytes[:])
+	contents, err = CppifyHeaderIncludes(contents)
+	if err != nil {
+		return err
+	}
+	b := []byte(contents)
+	err = ioutil.WriteFile(srcFile, b, 0644)
+	return err
+}
+
+func CppifyHeaderIncludes(contents string) (string, error) {
+	shouldCppify, exists := os.LookupEnv("JCC_CPPIFY_PROJECT_HEADERS")
+	if !exists || strings.Compare(shouldCppify, "0") == 0 {
+		return contents, nil
+	}
+	if strings.Contains(contents, CppifyHeadersMagicString) {
+		return contents, nil
+	}
+	re := regexp.MustCompile(`\#include \"(?P<header>.+)\"\n`)
+	matches := re.FindAllStringSubmatch(contents, -1)
+	if len(matches) == 0 {
+		return "", nil // !!!
+	}
+	for i, match := range matches {
+		if i == 0 {
+			// So we don't cppify twice.
+			contents += CppifyHeadersMagicString
+		}
+		oldStr := match[0]
+		replacement := "extern \"C\" {\n#include \"" + match[1] + "\"\n}\n"
+		contents = strings.Replace(contents, oldStr, replacement, 1)
+		if strings.Compare(contents, "") == 0 {
+			panic("Failed to replace")
+		}
+	}
+	return contents, nil
+}
+
+func AppendStringToFile(filepath, new_content string) error {
+	// Appends |new_content| to the content of |filepath|.
+	file, err := os.OpenFile(filepath, os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
Not a good idea to remove append here, it will affect other functions like `WriteStdErrOut` when appending err logs in `WriteStdErrOut`.